sycl-fusion/passes/kernel-fusion/SYCLSpecConstMaterializer.cpp

```diff
@@ -0,0 +1,317 @@
+      break;
+    }
+    }
+  }
+
+  const DataLayout &DL = Mod->getDataLayout();
+  for (auto *AS : ASCasts) {
+    for (auto *U : AS->users()) {
+      auto *I = cast<Instruction>(&*U);
+      switch (I->getOpcode()) {
+      default: {
+        MATERIALIZER_DEBUG(
+            dbgs()
+            << "Optimization opportunity missed, unhandled instruction: \n");
+        MATERIALIZER_DEBUG(I->dump());
+        MATERIALIZER_DEBUG(dbgs() << "Function:\n");
+        MATERIALIZER_DEBUG(I->getParent()->getParent()->dump());
+        break;
+      }
+      case Instruction::Load: {
+        TypeAtOffset TyO{
+            I->getType(),
+            /* Non GEP load starts at the beginnig of memory region */ 0};
+        TypesAndOffsets.insert(TyO);
+        LoadsToTypes[cast<LoadInst>(I)] = TyO;
+        break;
+      }
+      case Instruction::GetElementPtr: {
+        auto *GEP = cast<GetElementPtrInst>(I);
+        unsigned int ASL = GEP->getPointerAddressSpace();
+        unsigned OffsetBitWidth = DL.getIndexSizeInBits(ASL);
+        APInt Offset(OffsetBitWidth, 0);
+        bool FoundOffset = GEP->accumulateConstantOffset(DL, Offset);
+        if (!FoundOffset)
+          llvm_unreachable_internal("Offset unknown.");
+        auto Loads = collectGEPsLoads(GEP);
+        for (auto *Load : Loads) {
+          TypeAtOffset TyO{Load->getType(), Offset.getSExtValue()};
+          TypesAndOffsets.insert(TyO);
+          LoadsToTypes[Load] = TyO;
+        }
+        break;
+      }
+      }
+    }
+  }
+}
+
+void SYCLSpecConstMaterializer::reportAndReset() {
+  if (LoadsToTypes.empty()) {
+    MATERIALIZER_DEBUG(dbgs()
+                       << "Did not find any loads from spec const buffer.\n");
+  } else {
+    MATERIALIZER_DEBUG(dbgs() << "Replaced: " << LoadsToTypes.size()
+                              << " loads from spec const buffer.\n");
+    MATERIALIZER_DEBUG(dbgs() << "Load to global variable mappings:\n");
+    for (auto &LTT : LoadsToTypes) {
+      MATERIALIZER_DEBUG(dbgs() << "\tLoad:\n");
+      MATERIALIZER_DEBUG(LTT.first->dump());
+      MATERIALIZER_DEBUG(dbgs() << "\tGlobal Variable:\n");
+      MATERIALIZER_DEBUG(TypesAndOffsetsToBlob[LTT.second]->dump());
+      MATERIALIZER_DEBUG(dbgs() << "\n");
+    }
+  }
+  MATERIALIZER_DEBUG(dbgs() << "\n\n");
+
+  // Reset the state.
+  TypesAndOffsets.clear();
+  TypesAndOffsetsToBlob.clear();
+  LoadsToTypes.clear();
+}
+
+PreservedAnalyses
+SYCLSpecConstMaterializer::handleKernel(llvm::Function &Kernel) {
+  if (Kernel.arg_empty())
+    return PreservedAnalyses::all();
+  auto *SpecConstArg = std::prev(Kernel.arg_end());
+  if (!SpecConstArg || !SpecConstArg->hasName() ||
+      (SpecConstArg->getName() != "_arg__specialization_constants_buffer"))
+    return PreservedAnalyses::all();
+
+  if (!readMetadata())
+    return PreservedAnalyses::all();
+
+  if (!SpecConstData || SpecConstDataSize < 1)
+    llvm_unreachable("Specialisation constant data not found");
```

Comment:
Assert

Reply:
Why?

Reply:
That looks like an invariant to me, so we should be checking the invariant with an `assert`

Reply:
How does `llvm_unreachable` behave in builds with assertions disabled?

Reply:
It's a full on failure, so much so, that, if everything else fails, they [dereference `0x11`](https://github.com/jchlanda/llvm/blob/jakub/jit_spec_const/llvm/include/llvm/Support/Compiler.h#L375) to make sure it crushes and burns.

A nice thing about unreachable is that it is evaluated at release, so you don't need to do all that might be unused warnings in a single statement (`assert`) `if` clauses. I'm not married to it though, so don't mind changing it.

Reply:
Nah, `assert` it is. Your are right, it's an invariant more than an unreachable condition.